### PR TITLE
Allow shard pruning with IN literal

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):

- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
For distributed query, when `optimize_skip_unused_shards=1`, allow to skip shard with condition like `(sharding key) IN (one-element-tuple)`. (Tuples with many elements were supported. Tuple with single element did not work because it is parsed as literal).

### CODE REVIEW

**Review by KochetovNicolai — APPROVED** [review]

File: tests/queries/0_stateless/01756_optimize_skip_unused_shards_rewrite_in.sql [hunk 1]
```diff
@@ -86,13 +86,8 @@ order by query;
 --
 select 'errors';
 
--- not tuple
-select * from dist_01756 where dummy in (0); -- { serverError 507 }
```
Comment (azat) [inline tests/queries/0_stateless/01756_optimize_skip_unused_shards_rewrite_in.sql:90]: Can you expectations for them instead?

File: src/Storages/StorageDistributed.cpp [hunk 2]
```diff
@@ -609,7 +609,7 @@ void StorageDistributed::read(
     ClusterProxy::executeQuery(query_plan, select_stream_factory, log,
         modified_query_ast, local_context, query_info,
         sharding_key_expr, sharding_key_column_name,
-        getCluster());
+        query_info.cluster);
```
Comment (azat) [inline src/Storages/StorageDistributed.cpp:612]: Deserve a separate commit, since it fixes the race between `getQueryProcessingStage()` and `read()` (if the cluster definition will be changed in between).

File: tests/queries/0_stateless/01756_optimize_skip_unused_shards_rewrite_in.sql (hunk 1 above)
Comment (amosbird) [reply tests/queries/0_stateless/01756_optimize_skip_unused_shards_rewrite_in.sql:90]: done

File: src/Storages/StorageDistributed.cpp (hunk 2 above)
Comment (amosbird) [reply src/Storages/StorageDistributed.cpp:612]: done
